### DISCUSSION TITLE
How can I run a struct‘s method in tokio::spawn?

### DISCUSSION THREAD

**Discussion by fakecore** [tokio-rs/tokio#4686] [Q&A] [answered]
I want to use a struct to manage connected TcpStream.
e.g.
```rust
tokio::spawn(
                async move {
                    let mut ctx = Context::new(socket);
                    let mut service = NetService::new(&mut ctx);
                    service.handle_connected();
                    service.run().await;
                    service.handle_disconnected();
                }
            );
```
It failed to pass the compile, How do I modify the code to pass the compiler?

**Comment by Darksonn** [comment, marked as answer]
Without seeing the compiler error, I can't help you.

**Reply by fakecore** [reply]
Sorry for the late reply. 
this is the error log. 
I try to use the OOP style like a struct includes the biz functions.But it is failed to pass the compiler's check.
What should  I do to get the right code? 
```
error: future cannot be sent between threads safely
   --> src/net/server.rs:39:13
    |
39  |             task::spawn(
    |             ^^^^^^^^^^^ future created by async block is not `Send`
    |
    = help: the trait `std::marker::Send` is not implemented for `(dyn for<'r, 's> FnMut(&'r mut context::Context, &'s Vec<u8>) + 'static)`
note: future is not `Send` as this value is used across an await
   --> src/net/server.rs:44:34
    |
42  |                     let mut service = NetService::new(&mut ctx);
    |                         ----------- has type `NetService<'_>` which is not `Send`
43  |                     service.handle_connected();
44  |                     service.run().await;
    |                                  ^^^^^^ await occurs here, with `mut service` maybe used later
45  |                     service.handle_disconnected();
46  |                 }
    |                 - `mut service` is later dropped here
note: required by a bound in `tokio::spawn`
   --> /Users/dylan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/task/spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ^^^^ required by this bound in `tokio::spawn`


```

**Reply by Darksonn** [reply]
It seems like the issue is that you are storing a `Box<dyn FnMut(&mut Context, &Vec<u8>)>` somewhere. You should add `+ Send` to it so it looks like this:
```
Box<dyn FnMut(&mut Context, &Vec<u8>) + Send>
```

**Reply by fakecore** [reply]
> It seems like the issue is that you are storing a `Box<dyn FnMut(&mut Context, &Vec<u8>)>` somewhere. You should add `+ Send` to it so it looks like this:
> 
> ```
> Box<dyn FnMut(&mut Context, &Vec<u8>) + Send>
> ```

It works !!! thx, Darksonn.This problem bothered me few days, I should come here and ask early.
But why does it works? Could u pls tell me how it works or where I can some material about this?

**Reply by Darksonn** [reply]
A `Box<dyn FnMut(&mut Context, &Vec<u8>)>` without the `+ Send` _could_ contain a function that may not be sent across threads. Therefore, sending the box across threads is not allowed. However, this is a problem because things spawned with `tokio::spawn` may be moved from one thread to another at any `.await`.

The fact that the box does not _actually_ contain a non-Send function is irrelevant. It's type allows it to do so, and the compiler does not look through every place you construct the box to figure out whether it ever contains a non-Send function.

**Reply by fakecore** [reply]
Ok, I see.
I saw the documents of tokio and rust multithread programming. Also, I noticed that `+Send` is necessary for multithread programming. I still didn't know how to solve it.Because i didn't know the `+Send`  could use on  `Box<dyn FnMut(&mut Context, &Vec<u8>) + Send>`.
Commonly, If you want to use a trait, you must use such grammar as `impl trait for xx`.
So, It's an implicit transfer or implicit implement?

**Reply by Darksonn** [reply]
Whenever you see `dyn SomeTraitGoesHere`, this refers to a special type called a "trait object". It's different from the trait `SomeTraitGoesHere`, which is a trait rather than a type. The trait object type can contain values of any type that implements the trait. To expand on this, trait objects can require multiple traits with a plus, which is what's happening when we add `+ Send`. When you do this, the trait object can only contain values of types that implement both traits.

Important thing to note here is that without the `+ Send`, the trait object could contain functions that don't implement `Send`, so the trait object type cannot be `Send`.

As for whether `Send` is implicitly implemented, yes. It is called an _auto trait_, which means that it is implemented for a type when all of its fields implement it.

**Reply by fakecore** [reply]
Thank you. I had read the feature 'auto trait'.It helps me a lot to understand.
